src/vmhistogram.rs

```diff
@@ -0,0 +1,328 @@
+///
+/// See https://medium.com/@valyala/improving-histogram-usability-for-prometheus-and-grafana-bc7e5df0e350
+///
+/// Each bucket contains a counter for values in the given range.
+/// Each non-empty bucket is exposed via the following metric:
+///
+///	<metric_name>_bucket{<optional_tags>,vmrange="<start>...<end>"} <counter>
+///
+/// Where:
+///
+///   - <metric_name> is the metric name passed to NewHistogram
+///   - <optional_tags> is optional tags for the <metric_name>, which are passed to NewHistogram
+///   - <start> and <end> - start and end values for the given bucket
+///   - <counter> - the number of hits to the given bucket during Update* calls
+///
+/// Histogram buckets can be converted to Prometheus-like buckets with `le` labels
+/// with `prometheus_buckets(<metric_name>_bucket)` function from PromQL extensions in VictoriaMetrics.
+/// (see https://github.com/VictoriaMetrics/VictoriaMetrics/wiki/MetricsQL ):
+///
+///	prometheus_buckets(request_duration_bucket)
+///
+/// Time series produced by the Histogram have better compression ratio comparing to
+/// Prometheus histogram buckets with `le` labels, since they don't include counters
+/// for all the previous buckets.
+///
+/// Zero histogram is usable.
+#[allow(missing_debug_implementations)]
+#[derive(Clone)]
+pub struct VMHistogram {
+    inner: Arc<Mutex<Inner>>,
+    desc: Arc<Desc>,
+    label_pairs: Arc<Vec<LabelPair>>,
+}
+
+impl VMHistogram {
+    /// `with_opts` creates a [`VMHistogram`] with the `opts` options.
+    pub fn with_opts(opts: Opts) -> crate::Result<Self> {
+        VMHistogram::with_opts_and_label_values(&opts, &[])
+    }
+
+    fn with_opts_and_label_values(opts: &Opts, label_values: &[&str]) -> crate::Result<Self> {
+        let desc = opts.describe()?;
+
+        let label_pairs = make_label_pairs(&desc, label_values)?;
+        Ok(VMHistogram {
+            inner: Arc::new(Mutex::new(Inner::default())),
+            desc: Arc::new(desc),
+            label_pairs: Arc::new(label_pairs),
+        })
+    }
+
+    /// Add a single observation to the [`VMHistogram`]
+    /// Negative values and NaNs are ignored.
+    pub fn observe(&self, value: f64) {
+        if value.is_nan() || value.is_sign_negative() {
+            return;
+        }
+        let bucket_idx = (value.log10() - E_10_MIN as f64) * BUCKETS_PER_DECIMAL as f64;
+        let mut inner = self.inner.lock();
+        inner.sum += value;
+        if bucket_idx.is_sign_negative() {
+            inner.lower += 1;
+        } else if bucket_idx as usize >= BUCKETS_COUNT {
+            inner.upper += 1;
+        } else {
+            let mut idx = bucket_idx as usize;
+            if idx as f64 == bucket_idx && idx > 0 {
+                // Edge case for 10^n values, which must go to the lower bucket
+                // according to Prometheus logic for `le`-based histograms.
+                idx = idx - 1;
```

Comment:
```suggestion
                idx -= 1;
```

Reply:
🤦 I misremembered rust not having `--` and `++` as rust not having `-=` and `+=`